### PR TITLE
tiff exporter callback

### PR DESCRIPTION
**Pull request by chiahaoliu**
This PR includes a ``save_tiff`` in a `Callback`` version, so that it is more robust with errors which might happen during a scan.

@CJ-Wright I got stuck with ``pytest`` (limited experience), error message shows:

````IndexError: There are only 0 runs.````

 in my test. Could you give me some guides/hints once you get a chance? thanks.

### CODE REVIEW

File: xpdan/tests/test_tiff_exporter.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+template = '/direct/XF28ID1/pe2_data/xpdUser/tiff_base/{start.sample_name}'
+data_fields = ['temperature', 'diff_x', 'diff_y', 'eurotherm'] # known devices
+
+
+tif_export = XpdAcqLiveTiffExporter('pe1_image', template, data_fields,
+                                    overwrite=True, db=an_glbl.exp_db)
+#alias
+def test_tiff_export_smoke_test():
+    an_glbl.exp_db.process(an_glbl.exp_db[-1], tif_export)
```
Comment (CJ-Wright) [inline xpdan/tests/test_tiff_exporter.py]: Don't call the global db, just put `exp_db` in the signature of the function, also you should set up the config things in the test, or as a parametric test.
I think this might work better for you:
```python
def test_tiff_export_smoke_test(exp_db):
    tif_export = XpdAcqLiveTiffExporter('pe1_image', template, data_fields,
                                     overwrite=True, db=exp_db)
    exp_db.process(exp_db[-1], tif_export)
```

File: xpdan/callbacks_core.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        The databroker instance to use, if not provided use databroker
+        singleton
+    """
+
+    def __init__(self, field, data_dir_template,
+                 data_fields=[], save_dark=False,
+                 dryrun=False, overwrite=False, db=None):
+        try:
+            import tifffile
+        except ImportError:
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: I would just let this break. The trace back may give us a bit more info than just the print statement. Also Tifffile is not the only import which could break this. Also also, I expect this to be a minor issue, we will require tifffile for the conda build and I think it is a dep for most of the NSLS-II stack.

File: xpdan/callbacks_core.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    db : Broker, optional
+        The databroker instance to use, if not provided use databroker
+        singleton
+    """
+
+    def __init__(self, field, data_dir_template,
+                 data_fields=[], save_dark=False,
+                 dryrun=False, overwrite=False, db=None):
+        try:
+            import tifffile
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: Why is this here and not in the top?

File: xpdan/callbacks_core.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        except ImportError:
+            print("Tifffile is required by this callback. Please install"
+                  "tifffile and then try again."
+                  "\n\n\tpip install tifffile\n\nor\n\n\tconda install "
+                  "tifffile")
+            raise
+        else:
+            # stash a reference so the module is accessible in self._save_image
+            self._tifffile = tifffile
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: Why can't we just use the tifffile that we import?

File: xpdan/callbacks_core.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+        if db is None:
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: Can we just hand this in as a default?

File: xpdan/tests/test_tiff_exporter.py (hunk 1 above)
Comment (chiahaoliu) [inline xpdan/tests/test_tiff_exporter.py]: yup, working on it.

File: xpdan/callbacks_core.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+        return total_filename
+
+    def _save_image(self, image, filename):
+        """method to save image"""
+        dir_path, fn = os.path.split(filename)
+        os.makedirs(dir_path, exist_ok=True)
+
+        if not self.overwrite:
+            if os.path.isfile(filename):
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: can you compress this into one statement?
```python
if not self.overwrite and os.path.isfile(filename):
```

File: xpdan/callbacks_core.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            self._tifffile.imsave(filename, np.asarray(image))
+            print("INFO: {} has been saved at {}"
+                  .format(fn, dir_path))
+
+        self.filenames.append(filename)
+
+    def _pull_dark_uid(self, doc, dark_field_key='sc_dk_field_uid'):
+        """method to relate dark to images
+
+        Simply replace this method if scheme is changed in the future
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: We really should put a schema number into the header to check this.

File: xpdan/callbacks_core.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                                                         'sub_'+fn))
+                else:
+                    self._save_image(plane, filename)
+                # if user wants wants raw dark
+                if self.save_dark:
+                    self._save_image(self.dark_img, os.path.join(path_dir,
+                                                                 'dark_'+fn))
+    def stop(self, doc):
+        """method for stop document"""
+        # TODO: include sum logic in the future
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: What do you mean sum logic?

**Comment by CJ-Wright** [comment]
@chiahaoliu Can you please add matplotlib to the conda install section of the `travis.yml`?

**Comment by chiahaoliu** [comment]
sure, I will add it

**Comment by CJ-Wright** [comment]
Awesome, thanks! I think that that will help with the matplotlib issue you are having in he CI

File: xpdan/callbacks_core.py (hunk 8 above)
Comment (chiahaoliu) [inline xpdan/callbacks_core.py]: you image summing logic

File: xpdan/callbacks_core.py (hunk 4 above)
Comment (chiahaoliu) [inline xpdan/callbacks_core.py]: fix in commit

File: xpdan/callbacks_core.py (hunk 2 above)
Comment (chiahaoliu) [inline xpdan/callbacks_core.py]: fix in commit

**Comment by chiahaoliu** [comment]
@CJ-Wright  updated PR is pushed. Though I still have ``matplotlib`` error after adding it to conda build section. Do you have any idea about this?

Also, while I am writing test, I realize we might need another PR to import ``insert_img`` (adding more fields in ``data`` document and  other possible fields) so that we can have a more robust test.

File: xpdan/callbacks_core.py (hunk 8 above)
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: Can we do that in a separate callback?

**Comment by CJ-Wright** [comment]
Hmm, we could install bluesky using conda and maybe that will fix it. @tacaswell any thoughts? We are getting the following error:
```python
ImportError while importing test module '/home/travis/build/xpdAcq/xpdAn/xpdan/tests/test_tiff_exporter.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../mc/envs/testenv/lib/python3.5/site-packages/_pytest/python.py:415: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
../../../mc/envs/testenv/lib/python3.5/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
../../../mc/envs/testenv/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:212: in load_module
    py.builtin.exec_(co, mod.__dict__)
xpdan/tests/test_tiff_exporter.py:5: in <module>
    from xpdan.callbacks_core import XpdAcqLiveTiffExporter
xpdan/callbacks_core.py:6: in <module>
    from bluesky.callbacks.core import CallbackBase
../../../mc/envs/testenv/lib/python3.5/site-packages/bluesky/callbacks/__init__.py:1: in <module>
    from .core import (CallbackBase, CallbackCounter, print_metadata, collector,
../../../mc/envs/testenv/lib/python3.5/site-packages/bluesky/callbacks/core.py:9: in <module>
    import matplotlib.pyplot as plt
../../../mc/envs/testenv/lib/python3.5/site-packages/matplotlib/pyplot.py:114: in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
../../../mc/envs/testenv/lib/python3.5/site-packages/matplotlib/backends/__init__.py:32: in pylab_setup
    globals(),locals(),[backend_name],0)
../../../mc/envs/testenv/lib/python3.5/site-packages/matplotlib/backends/backend_qt4agg.py:18: in <module>
    from .backend_qt5agg import FigureCanvasQTAggBase as _FigureCanvasQTAggBase
../../../mc/envs/testenv/lib/python3.5/site-packages/matplotlib/backends/backend_qt5agg.py:16: in <module>
    from .backend_qt5 import QtCore
../../../mc/envs/testenv/lib/python3.5/site-packages/matplotlib/backends/backend_qt5.py:31: in <module>
    from .qt_compat import QtCore, QtGui, QtWidgets, _getSaveFileName, __version__
../../../mc/envs/testenv/lib/python3.5/site-packages/matplotlib/backends/qt_compat.py:137: in <module>
    from PyQt4 import QtCore, QtGui
E   ImportError: libicui18n.so.56: cannot open shared object file: No such file or directory
!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
```

File: xpdan/callbacks_core.py (hunk 8 above)
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: It might be better if we could separate them if possible.

File: xpdan/callbacks_core.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            if self._find_dark:
+                self._save_image(image, os.path.join(path_dir,
+                                                     'sub_'+fn))
+            else:
+                self._save_image(image, filename)
+            # if user wants wants raw dark
+            if self.save_dark:
+                self._save_image(self.dark_img, os.path.join(path_dir,
+                                                             'dark_'+fn))
+        if image.ndim == 3:
```
Comment (CJ-Wright) [inline xpdan/callbacks_core.py]: Would it be possible to make an function which does this (if it doesn't exist already) and only operates on 2D arrays? That way if the array is 3D we can just call the 2D one a bunch of times? That might help us to avoid reusing code

File: xpdan/tests/test_tiff_exporter.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import os
+import tempfile
+from tifffile import imread
+from xpdan.tests.conftest import img_size
+from xpdan.callbacks_core import XpdAcqLiveTiffExporter
+
+# standard config
+base = tempfile.mkdtemp()
+template = os.path.join(base, 'xpdUser/tiff_base/')
```
Comment (CJ-Wright) [inline xpdan/tests/test_tiff_exporter.py]: Can you use a pytest fixture for this? That way after the test is finished the folder can be removed.

File: xpdan/callbacks_core.py (hunk 8 above)
Comment (chiahaoliu) [inline xpdan/callbacks_core.py]: yeah, I am not intending to include them into the same PR either.

File: xpdan/tests/test_tiff_exporter.py (hunk 10 above)
Comment (chiahaoliu) [inline xpdan/tests/test_tiff_exporter.py]: working on it.

File: xpdan/callbacks_core.py (hunk 9 above)
Comment (chiahaoliu) [inline xpdan/callbacks_core.py]: I ended up extending the dimension of image array so 2D array with one frame would have shape as ``(1, x_dim, y_dim)`` so that we can always process as 3D array.

**Comment by chiahaoliu** [comment]
@CJ-Wright I rewrote the testing code with ``pytest`` fixture. Can you take a look very quick thanks?

**Comment by CJ-Wright** [comment]
looks reasonable to me

**Comment by CJ-Wright** [comment]
Unless anyone has objections I am going to merge this so I can put it into the pipeline for people wanting a file based workflow (I most likely will tweek it a little as we are doing the dark subtraction in a separate step) @chiahaoliu @sbillinge
